File: cpp/Sort/insertSort.cpp
#include "Sort.h"

int Sort::insertSort(vector<int> &nums, bool asc) {
	for (int i = 0; i < nums.size(); i++) {
		int tmp = nums[i];
		int j = i - 1;
		for (; j >= 0 && tmp < nums[j]; j--) nums[j + 1] = nums[j];
		nums[j + 1] = tmp;
	}
	if (!asc) Sort::reverse(nums);
	return 0;
}

File: js/Sort/insertSort.js
function swap(arr, i, j) {
    var tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
    return arr
}

function reverse(nums) {
    for (var i=0; i<parseInt(nums.length / 2); i++) nums = swap(nums, i, nums.length-1-i);
    return nums;
}

function insertSort(nums, asc) {
    for (var i=1; i<nums.length; i++) {
        var tmp = nums[i];
        for (var j=i-1; j>=0 && nums[j] > tmp; j--) {
            nums = swap(nums, j, j+1);
        }
        nums[j+1] = tmp;
    }
    if (!asc) nums = reverse(nums);
    return nums;
}

var res = insertSort([2,4,1,2], false);
console.log(res);